### DISCUSSION TITLE
Samletråd for React 18

### DISCUSSION THREAD

**Discussion by wkillerud** [fremtind/jokul#2824] [Annet]
Lager en samlepost her så vi kan hjelpe hverandre med oppgraderingen til [React 18](https://reactjs.org/blog/2022/03/29/react-v18.html). Relatert PR: #2822 

- Har du prøvd å oppgradere? 
- Hvilke problemer støtte du på, og med hvilke bibliotek? 
- Er det noe i Jøkul som blocker deg?
- Noen tips og triks du vil dele som gjorde ting enklere for deg?
- Rants? 🤬

Go Team!

![](https://media.giphy.com/media/26xBAevSH9ORZM6tO/giphy.gif)

**Comment by wkillerud** [comment]
Jeg fikk god nytte av å legge til [`<StrictMode>`](https://reactjs.org/docs/strict-mode.html#ensuring-reusable-state) for å luke ut et par sære bugs i `useAnimatedHeight` og `DatePicker`. Det er endringer i måten komponenter kan rendres på i v18, og vi ble truffet av det et par steder i Jøkul. Det var to steder hvor vi brukte en `ref` for å prøve å holde styr på om dette var komponentens første render, og så styre logikk ut fra det. Det er _no bueno_ i en concurrent render-verden🙅‍♂️

**Reply by wkillerud** [reply]
@Saegrov, jeg tror [dette er grunnen](https://github.com/fremtind/jokul/pull/2822/commits/1941cd8eae3e180915e7371755610c34098414e1) til den bugen du så i Select og [Datepicker](https://github.com/fremtind/jokul/pull/2822/commits/4e8ef6ad8edb284061fe990cd5272eab2226a02e)

**Comment by wkillerud** [comment]
Vi bruker `testing-libary/user-event` for å simulere klikk, men den nyeste versjonen ser ut til å ha en bug i måten den simulerer på som treffer oss i Jøkul i hvertfall. Den setter seg fast på ett av eventene før klikket, men bare for noen av knappene våre (testen feiler pga timeout). 

Jeg skal se om jeg får isolert problemet, men en workaround er å sende inn option `skipHover`. For øvrig verdt å få med seg at [siden v14 er alle funksjonene gjort asynkrone](https://github.com/testing-library/user-event/releases/tag/v14.0.0).

```js
await userEvent.click(button, { skipHover: true });
```

**Comment by wkillerud** [comment]
Vi bruker også `testing-library/react-hooks` for å teste noen av hooksene våre utenfor komponent-context. Der ser det ut som veien videre er at `renderHook` flyttes til `testing-library/react` (er en åpen PR, nr 991 – dropper lenke for å unngå støy). Det er ikke helt klart ennå, men det funker fremdeles å bruke dagens versjon av `testing-library/react-hooks`. Du ender bare opp med mange stygge warnings i konsollen.

**Reply by wkillerud** [reply]
`renderHook` skal nå være tilgjengelig via `testing-library/react` :) Sjekk PRen jeg nevner over om du vil ha _the deets_.

**Comment by wkillerud** [comment]
Edit: [Typene for React 18 er ute!](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56210) Finnes codemods å prøve for å fikse eventuelle feil. Skal få testet i Jøkul, men blir nok først etter påska.

`@types/react` og `@types/react-com` er fremdeles på v17, så TS vil klage om du ønsker å bruke nye API som f. eks. `useId`. PR nr 56210  i DefinitelyTyped er under utvikling. Enn så lenge må vi skru av typesjekking for de nye APIene eller lage våre egne midlertidige.

**Reply by wkillerud** [reply]
Stryk det: https://blog.logrocket.com/how-to-use-typescript-with-react-18-alpha/

tl;dr - legg til `"types": ["react/next", "react-dom/next"]` i `tsconfig.json`

Edit: det [krever at man definerer _alle_ types](https://www.typescriptlang.org/tsconfig#types) man trenger. Ikke helt noe jeg har lyst til, så her blir det brått `// @ts-ignore: wait for types to land in main`

**Reply by wkillerud** [reply]
Dere som oppdaterer typer oppdager nok fort at ting brekker. Denne bloggposten har en oppsummering av hva som knekker og hvorfor det er gjort: https://solverfox.dev/writing/no-implicit-children/

**Comment by wkillerud** [comment]
> Hydration failed because the initial UI does not match what was rendered on the server.

Denne treffer oss flere steder i portalen, og vil nok treffe ute i appene også. [The Perils of Rehydration](https://www.joshwcomeau.com/react/the-perils-of-rehydration/) forklarer godt hva som skjer, hvorfor, og en workaround.

tl;dr (men anbefaler å lese for å forstå hvorfor!) - unngå `typeof window !== "undefined"`-sjekker i komponenter og hooks, bruk et pattern som dette i stedet:

```tsx
  const [hasMounted, setHasMounted] = useState(false);
  useEffect(() => {
    setHasMounted(true);
  }, []);
  if (!hasMounted) {
    return null;
  }
```

**Reply by wkillerud** [reply]
Jeg vil tro det ikke trengs der man bare legger til og fjerner event handlers. Dette er mest for de stedene vi bruker noe fra `window` for å bestemme hva som skal rendres, for eksempel med `useScreen` og `useBrowserPreferences`.

**Comment by joms** [comment]
Hvordan vil det fungere å ta i bruk `useId`? Vil det brekke kompitabilitet? Den hadde vært veldig praktisk i server-rendrete applikasjoner, hvor f.eks. `Logo` gir feilmelding på at id ikke matcher mellom server og klient.

**Reply by wkillerud** [reply]
Jeg prøver meg fram på #2822 (https://github.com/fremtind/jokul/pull/2822/commits/4845325b962224e012335fcfff568da3d9cab7ed konkret).

**Comment by kennidenni** [comment]
Møtte på et problem med at jest ikke klarte å bruke nanoId som brukes i useId.ts fra Jøkul.
Måtte legge til denne koden inne i jest.config.js: 
`    
moduleNameMapper: {
        "^nanoid$": require.resolve("nanoid"),
    },
`
